bee-protocol/src/worker/transaction/transaction.rs

```diff
@@ -113,51 +114,60 @@ impl TransactionWorker {
         }
 
-        match tangle().insert_transaction(transaction, hash).await {
-            Some(transaction) => {
-                Protocol::get().metrics.new_transactions_received_inc();
-                if !tangle().is_synced() && Protocol::get().requested.is_empty() {
-                    Protocol::trigger_milestone_solidification().await;
+        // store transaction
+        if let Some(transaction) = tangle().insert(transaction, hash, Flags::empty()) {
+            Protocol::get().metrics.new_transactions_received_inc();
+
+            if !tangle().is_synced() && Protocol::get().requested.is_empty() {
+                Protocol::trigger_milestone_solidification().await;
+            }
+
+            match Protocol::get().requested.remove(&hash) {
+                Some((hash, index)) => {
+                    Protocol::trigger_transaction_solidification(hash, index).await;
                 }
-                match Protocol::get().requested.remove(&hash) {
-                    Some((hash, index)) => {
-                        Protocol::trigger_transaction_solidification(hash, index).await;
-                    }
-                    None => Protocol::broadcast_transaction_message(Some(from), transaction_broadcast).await,
-                };
+                None => Protocol::broadcast_transaction_message(Some(from), transaction_broadcast).await,
+            };
 
-                if transaction.address().eq(&Protocol::get().config.coordinator.public_key)
-                    || transaction.address().eq(&Protocol::get().config.null_address)
-                {
-                    let tail = {
-                        if transaction.is_tail() {
-                            Some(hash)
-                        } else {
-                            let chain =
-                                tangle().trunk_walk_approvers(hash, |tx_ref| tx_ref.bundle() == transaction.bundle());
-                            match chain.last() {
-                                Some((tx_ref, hash)) => {
-                                    if tx_ref.is_tail() {
-                                        Some(*hash)
-                                    } else {
-                                        None
-                                    }
+            if transaction.address().eq(&Protocol::get().config.coordinator.public_key)
+                || transaction.address().eq(&Protocol::get().config.null_address)
+            {
+                let tail = {
+                    if transaction.is_tail() {
+                        Some(hash)
+                    } else {
+                        let mut bundle = vec![];
+
+                        traversal::visit_children_follow_trunk(
+                            &tangle().inner,
+                            hash,
+                            |tx, _| tx.bundle() == transaction.bundle(),
+                            |tx_hash, tx, _| {
+                                bundle.push((*tx_hash, tx.clone()));
```

Comment:
Why are we storing in a bundle if we only need the last ?

Reply:
I don't know. It's not my code.

Reply:
Replaced it with `Option<(TxHash, TransactionRef)>`